crates/stages/src/stages/hashing_storage.rs

```diff
@@ -57,77 +95,92 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
+                self.save_checkpoint(tx, checkpoint)?;
+            }
 
-            let mut current_key = None;
-            let mut current_subkey = None;
+            let mut current_key = checkpoint.address.take();
+            let mut current_subkey = checkpoint.storage.take();
             let mut keccak_address = None;
 
-            loop {
-                let mut hashed_batch = BTreeMap::new();
-                let mut remaining = self.commit_threshold as usize;
-                {
-                    let mut storage = tx.cursor_dup_read::<tables::PlainStorageState>()?;
-                    while !remaining.is_zero() {
-                        hashed_batch.extend(
-                            storage
-                                .walk_dup(current_key, current_subkey)?
-                                .take(remaining)
-                                .map(|res| {
-                                    res.map(|(address, slot)| {
-                                        // Address caching for the first iteration when current_key
-                                        // is None
-                                        let keccak_address =
-                                            if let Some(keccak_address) = keccak_address {
-                                                keccak_address
-                                            } else {
-                                                keccak256(address)
-                                            };
-
-                                        // TODO cache map keccak256(slot.key) ?
-                                        ((keccak_address, keccak256(slot.key)), slot.value)
-                                    })
+            let mut hashed_batch = BTreeMap::new();
+            let mut remaining = self.commit_threshold as usize;
+            {
+                let mut storage = tx.cursor_dup_read::<tables::PlainStorageState>()?;
+                while !remaining.is_zero() {
+                    hashed_batch.extend(
+                        storage
+                            .walk_dup(current_key, current_subkey)?
+                            .take(remaining)
+                            .map(|res| {
+                                res.map(|(address, slot)| {
+                                    // Address caching for the first iteration when current_key
+                                    // is None
+                                    let keccak_address =
+                                        if let Some(keccak_address) = keccak_address {
+                                            keccak_address
+                                        } else {
+                                            keccak256(address)
+                                        };
+
+                                    // TODO cache map keccak256(slot.key) ?
+                                    ((keccak_address, keccak256(slot.key)), slot.value)
```

Comment:
yeah probably right solution here is just to have a LRU maps for map<address, keccak256(address)> and map<h256, keccak256(h256)> to avoid redundant hashing, but doubt it'll be a bottleneck